### ISSUE TITLE
Wrong scale for predictions from "predict(...,newdata=...)"

### ISSUE THREAD

**Issue by apollostream** [haziqjamil/iprior#45]
Great package! 

However, I'm getting odd predictions that appear to be on a different scale than the training data when specifying an "se" kernel and predicting with "newdata". I've attached a small R script with an example. (I changed the file extension to .txt to stop Github from complaining when I drag & drop.)

I hope this has an easy work-around. If you have one, please let me know.
Thank you.
[iprior_bug.txt](https://github.com/haziqj/iprior/files/7062384/iprior_bug.txt)

**Comment by haziqj** [comment]
Thanks! I'll have a look at that R script today, but I think I know what you're referring to. I remember encountering this on a real data set I was working on a few months ago, and I thought it was a bit odd at the time and chalked it up to some peculiarities with that data set.

If it's something I can fix I'll definitely do it. Thanks for letting me know.

**Comment by haziqj** [comment]
For posterity, reproduce the issue here.

```r
library(magrittr)
library(tidyverse)
library(iprior)

# Create some training data
set.seed(23)
N_train <- 13
train_data <- tibble(
  x = runif(N_train, 0, 8),
  y = 0.8 * sin(x) + rnorm(N_train, sd = 0.1)
)

# Fit an "SE" kernel iprior model
ip_mod <- iprior( 
  y ~ x, 
  data   = train_data, 
  kernel = "se",
  est.lengthscale = TRUE,
  method = "mixed"
)

# What about predicting exactly at the training data x values?
yf_obj <- fitted(ip_mod)
yptrain_obj <- predict(ip_mod, newdata = train_data )
fitted_data <- train_data %>%
  rename(Train = "y") %>%
  mutate(Fitted = yf_obj$y, Predicted = yptrain_obj$y ) %>%
  pivot_longer(cols = -x, names_to = "Source", values_to = "y")

# Plot the predictions and the data
fitted_data %>%
  {
    ggplot(.,aes(x=x,y=y)) +
      geom_point(aes(color = Source, alpha = Source, size = Source)) +
      scale_color_manual(values = c(Train = "black", Fitted = "red", Predicted = "green")) +
      scale_size_manual(values = c(Train = 3, Fitted = 3, Predicted = 5)) +
      scale_alpha_manual(values = c(Train = 1, Fitted = 1, Predicted = 0.4)) +
      labs( 
        title = "Fitted & Predicted at Training Input Values", 
        subtitle = "Works well."
      )
  } %>%
  print()
```

![image](https://user-images.githubusercontent.com/16189362/131072796-c4d615f1-215c-4ab1-95de-1d2d9e7fb942.png)

```r
# *** BUG APPEARS HERE: How about predicting at new points?
# Predict with the model at new points (many more for smooth plotting).
N_pred <- 100 # NOTE: N_pred > N_train
pred_data <- tibble(
  x = seq(min(train_data$x), max(train_data$x), length.out = N_pred)
)
yp_obj <- predict(ip_mod, newdata = pred_data, intervals = TRUE)
print(yp_obj) # Values appear to be too big!
pred_data %<>% 
  mutate(
    y     = yp_obj$y,
    lower = yp_obj$lower,
    upper = yp_obj$upper
  )

# Plot the predictions and the data.
pred_data %>%
  {
    ggplot(., aes(x = x, y = y)) +
      geom_line(size = 1, color = "blue") +
      geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.2, fill= "blue") +
      geom_point(data = train_data, size = 3 ) +
      labs( 
        title = "Predictions at Different X Values than Training X Values", 
        subtitle = "What! The curve completely misses the data!",
        caption = paste(
          "(Even tried this with N_pred = N_train",
          "and still seems to have small offset.)",
          sep = "\n"
        )
      )
  } %>%
  print()
```

![image](https://user-images.githubusercontent.com/16189362/131073031-2b309f74-1e70-4f14-90c9-9e88411e6d2b.png)

**Comment by haziqj** [comment]
TL;DR--using the un-centred SE kernel resolves the issue. I suspect the SE kernel has some properties which doesn't require centring, but that needs to be studied further.

------

Looking at the last graph above, it seems it is more a location issue rather than a scaling issue. I.e., the curve seems to be shifted by some amount, yet still looks like the correct shape.

This effect is not present when using fBm, linear, or polynomial kernels (although the latter two give an understandably horrible fit). This leads me to think that there is some peculiarity about the centring of the SE kernel. We studied the properties of centred canonical, fBm, Pearson and ANOVA RKHSs extensively, but not the SE RKHS. The motivation for centring RKHSs is because the origin of a Hilbert space may be arbitrary, and thus the location of the functions within it may not be where we're looking for it. Centring resolves this, because it yields, at least in the RKHSs we studied, zero-mean functions. This is nice, because once we find the regression function, we just have to shift it by the estimate of the intercept.

For some reason that's not yet clear to me, when the `centre` option is turned off for `kern_se()`, it now works. Going forward, the default behaviour for the SE kernel is to not centre. Maybe there's some properties of the SE kernel which I overlooked, that doesn't require centring--I'm not too sure. We prefer the fBm kernel instead of the SE. In many of our simulations and real-world applications, it seems to give an advantage over the smooth SE kernel in terms of prediction.

![image](https://user-images.githubusercontent.com/16189362/131074982-b43365e4-cf18-4585-8f97-69c972f37eb6.png)

**Comment by haziqj** [comment]
Thanks for bringing this up! I liked how the issue was emphasised with "What! The curve completely misses the data!"

**Comment by apollostream** [comment]
Excellent response.  Much appreciated. Thank you so much!
